proto/schema/v1/schema.proto

```diff
@@ -0,0 +1,41 @@
+syntax = "proto3";
+
+package schema.v1;
```

Comment:
I'm not sure if this warrants a whole independent package. I'd argue this is part of the interface between Conduit and the connector, so we could put it into `connector.v2`. Let me explain.

Let's think about the backwards compatibility that I added in https://github.com/ConduitIO/conduit-connector-protocol/pull/90. Updated connectors that use protocol v2 will work with older versions of Conduit, as they can fall back to protocol v1. The problem is that new connectors might also require this schema service, so they won't work correctly with older versions of Conduit. So maybe we should remove the backwards compatibility? 🤔 And this same reason is what makes me think that this belongs in `connector.v2`, as it makes it clear that the schema service is only available to connectors that use the connector protocol v2.

As an alternative, we could also put it into a "connectors utilities" package, where we can add more utilities in the future (like storage). This would mean that if we evolve the utilities and introduce a breaking change there, we don't have to necessarily bump the connector proto.

Reply:
@lovromazgon my first question is why to tie it to the connectors package at all? Maybe these could still be separate packages and say that connector.v2 will be compatible with schema.v1 being that the main reason to maybe upgrade to connector.v2 if they're still on connector.v1.

What do you think?

Reply:
A question first: what is `connector.v2` supposed to tell us? Is it v2 of the connector API, or is the connector part of protocol v2? I was assuming it's the first one (version 2 of connector API), because if we were versioning the whole protocol, I'd expect a structure like this:
```
# v1 of the protocol supports connectors and foos
v1
  - connector
  - foo
# in v2 of the protocol we introduced bar
v2:
  - connector
  - foo
  - bar
```

Also, my understanding of the `connector` folder is that it contains definitions of connector services, not necessarily everything a connector might use.

Based on the above assumptions I made the decision to create a schema.v1 package. I wouldn't like the schema service to be "too close" to the connector definitions, because they can evolve for different reasons and at different paces. 

Generally, I think putting it into a "connector utilities" is a good thing too. The only thing I don't like about it is that it gives the impression that schemas are for connectors only. Schemas as a feature will be used in processors, and if we had gRPC processors those would use it too.

Reply:
> Generally, I think putting it into a "connector utilities" is a good thing too. The only thing I don't like about it is that it gives the impression that schemas are for connectors only. Schemas as a feature will be used in processors, and if we had gRPC processors those would use it too.

Let's address this part first, so we're on the same page. We are in the `conduit-connector-protocol` repo, with emphasis on `connector` 🙃 As the readme says:

  _This repository is the only connection point between **Conduit** and a **connector**._

Stuff in this repo should _only_ be relevant to connectors, not processors, not the public API. If/when we expose schemas to processors, I'll argue that we should do it through WASM host functions, since that's the mechanism we use to communicate between a processor and Conduit. It would be a bit strange if a processor, running in the same process as Conduit, would communicate with Conduit through gRPC. As for the public API - that one has a different purpose and might need to be compatible with the Confluent API, if/when we decide to expose it to end-users.

> what is `connector.v2` supposed to tell us?

I see it as the definition of the interface between Conduit and the connector. And by interface I was thinking both ways - what Conduit exposes, and what the connector exposes.

But maybe we could say that `connector.v2` defines the services that the connector implements and we put this schema service into the package `conduit.v1`, since it's a service that Conduit exposes. So connector -> `connector.v2`, Conduit -> `conduit.v1`, both can evolve separately.

As for the versioning, we've been following the `FILE` protobuf [breaking rules](https://buf.build/docs/breaking/rules) (we have a CI action that checks this). It's completely valid and _not_ a breaking change if a new `rpc` gets added to the service. For instance, we did this with lifecycle events in the past. Conduit executes the lifecycle functions _if they exist_, otherwise it ignores them, as it's apparently an older connector. The problem appears if the opposite happens, e.g. we add a new `rpc` to the Conduit utilities service. That change is backwards compatible to old consumers (e.g. old plugins will still work fine, they don't care about the added `rpc`), but new consumers that start using the added `rpc` _will not_ be backwards compatible anymore, since they won't find the `rpc` in an older version of the service.

Why do we care? Because we don't want to find out a connector doesn't work when we start a pipeline and it suddenly breaks when it tries to hit an `rpc` that doesn't exist. Ideally, we should catch this early, e.g. when an older version of Conduit tries to register a connector, that expects the new version of a utility service, it should already fail to be registered.

I think we won't get around some kind of "handshake" between the connector and Conduit when the plugin gets registered, where the plugin can reject being registered if Conduit doesn't expose the expected services/rpcs. As mentioned in my previous comment, we could bump the major version of the whole proto package (e.g. `conduit.v1` -> `conduit.v2`), but that sounds wrong and overkill for changes that are essentially backwards compatible from the perspective of old consumers.

Reply:
Thanks for the detailed comment, Lovro! It cleared up some things. 

I was always reading the repo and its purpose as "a Conduit connector's protocol". And the protocol (services) defined in this repo are the only way to communicate with a connector. I was always reasoning about Conduit and a connector as being like very distinct (micro-)services, each with its own API. All (and only) the APIs Conduit exposes are defined in the Conduit repo, all (and only) connector APIs are in this repo. Anyway...

With things being as mentioned above, having something like `conduit.v1` makes sense to me.

Also, for the record: on the standup today, we discussed failing as early as possible when an older version of Conduit tries to load a newer version of a connector. While it's not a must-have for the schema support epic, it's still a very good improvement of user experience, and we'd like to have it. For now, we're leaning towards a solution that will be discussed in more details separately.

Reply:
Also, for the record: on the standup today, we discussed failing as early as possible when an older version of Conduit tries to load a newer version of a connector. While it's not a must-have for the schema support epic, it's still a very good improvement of user experience, and we'd like to have it. For now, we're leaning towards a solution that will be discussed in more details separately.

Reply:
I've been thinking more about this and I think this could be addressed by expanding the specification. The specification would include a list of services and their versions, so Conduit itself could check if it provides everything the connector expects when it fetches the specifications (this happens when registering the plugin). An additional benefit is that if we extract the [connector specifications into a YAML file](https://github.com/ConduitIO/conduit/discussions/1522), we can know if the connector is compatible with a specific Conduit version only by looking into the specification and we can display a compatibility matrix (or minimum Conduit version) in the docs.

So in the end, I think this can be safely added afterwards, all we need is a way for [specgen](https://github.com/ConduitIO/conduit/discussions/1522), to figure out what utilities the connector is using, then it can include those dependencies in the specifications yaml, everything else can happen in Conduit. This can still be done using the approach we discussed, by putting the utilities for a specific service into a separate package in the SDK.

Reply:
That sounds like a good idea! :+1:  It's similar to what we were discussing earlier about features.